rllib/core/rl_module/torch/torch_rl_module.py

```diff
@@ -1,25 +1,196 @@
+                backend=self.torch_dynamo_backend,
+                mode=self.torch_dynamo_mode,
+                **self.kwargs
+            )
+        if self.compile_forward_exploration:
+            rl_module.compile_forward_exploration(
+                backend=self.torch_dynamo_backend,
+                mode=self.torch_dynamo_mode,
+                **self.kwargs
+            )
+
+
 class TorchRLModule(nn.Module, RLModule):
+    """A base class for RLlib torch RLModules.
+
+    Note that the `_forward` methods of this class can be 'torch.compiled' individually:
+        - TorchRLModule._forward_train()
+        - TorchRLModule._forward_inference()
+        - TorchRLModule._forward_exploration()
+    As a rule of thumb, they should only contain torch-native tensor manipulations, or
+    otherwise they may yield wrong outputs. In particular, the creation of RLlib
+    distributions inside these methods should be avoided when using torch.compile.
+    When in doubt, you can use torch.dynamo.explain() to check whether a compiled
+    method has broken up into multiple sub-graphs.
+    Compiling these methods can bring speedups under certain conditions.
+    """
+
     framwork: str = "torch"
 
     def __init__(self, *args, **kwargs) -> None:
         nn.Module.__init__(self)
         RLModule.__init__(self, *args, **kwargs)
 
+        # Whether to retrace torch compiled forward methods on set_weights.
+        self._retrace_on_set_weights = False
+
+    @override(RLModule)
+    @check_input_specs("_input_specs_inference")
+    @check_output_specs("_output_specs_inference")
+    def forward_inference(self, batch: SampleBatchType, **kwargs) -> Mapping[str, Any]:
+        # If this forward method was compiled, we call the compiled version.
+        if hasattr(self, "_compiled_forward_inference"):
+            return self._compiled_forward_inference(batch, **kwargs)
+        return self._forward_inference(batch, **kwargs)
+
+    @override(RLModule)
+    @check_input_specs("_input_specs_exploration")
+    @check_output_specs("_output_specs_exploration")
+    def forward_exploration(
+        self, batch: SampleBatchType, **kwargs
+    ) -> Mapping[str, Any]:
+        # If this forward method was compiled, we call the compiled version.
+        if hasattr(self, "_compiled_forward_exploration"):
+            return self._compiled_forward_exploration(batch, **kwargs)
+        return self._forward_exploration(batch, **kwargs)
+
+    @override(RLModule)
+    @check_input_specs("_input_specs_train")
+    @check_output_specs("_output_specs_train")
+    def forward_train(self, batch: SampleBatchType, **kwargs) -> Mapping[str, Any]:
+        # If this forward method was compiled, we call the compiled version.
+        if hasattr(self, "_compiled_forward_train"):
+            return self._compiled_forward_train(batch, **kwargs)
+        return self._forward_train(batch, **kwargs)
+
+    def compile_forward_train(
```

Comment:
Oh yeah that's neat! Thanks

Reply:
@kouroshHakha How would you use this here? The decorator wraps the method at instantiation time, but we compile after instantiation time.

Reply:
based on the compile flag you could call something like compile_wrapper(module) and it would compile it using the same procedure outlined in the link that I sent?